lib/js/src/manager/screen/utils/SoftButtonObject.js

```diff
@@ -239,6 +243,144 @@ class SoftButtonObject {
+        currentState.setHighlighted(JSON.parse(JSON.stringify(this.getCurrentState().getHighlighted())));
+
+        if (this.getCurrentState().getSystemAction() !== null && this.getCurrentState().getSystemAction() !== undefined) {
+            currentState.setSystemAction(JSON.parse(JSON.stringify(this.getCurrentState().getSystemAction())));
+        }
+        const listenerClone = (this.getOnEventListener() !== undefined && this.getOnEventListener() !== null) ? this.getOnEventListener().bind({}) : null;
+        const clone = new SoftButtonObject(
+            JSON.parse(JSON.stringify(this.getName())),
+            statesClone,
+            currentState.getName(),
+            listenerClone
+        );
+        if (this._updateListener !== null && this._updateListener !== undefined) {
+            clone._setUpdateListener(this._updateListener.bind(clone));
+        }
+
+        clone._buttonId = JSON.parse(JSON.stringify(this.getButtonId()));
+        return clone;
+    }
+
+    /**
+     * Checks whether two SoftButtonObjects can be considered equivalent, but does NOT compare the listener objects
+     * @param {SoftButtonObject} other - The object to compare
+     * @returns {Boolean} - Whether the objects are the same or not
+     */
+    equals (other) {
+        if (other === null || other === undefined) {
+            return false;
+        }
+        if (this === other) {
+            return true;
+        }
+        if (!(other instanceof SoftButtonObject)) {
+            return false;
+        }
+        if (this.getName() !== other.getName()) {
+            return false;
+        }
+        const states = this.getStates();
+        const otherStates = other.getStates();
+
+        if ((states !== null && otherStates === null) || (states === null && otherStates !== null)) {
+            return false;
+        }
+        if ((states !== null && otherStates !== null)) {
+            if (states.length !== otherStates.length) {
+                return false;
+            }
+            for (let index = 0; index < states.length; index++) {
+                if (states[index] === null && otherStates[index] === null) {
+                    continue;
+                }
+                if (states[index].getName() !== otherStates[index].getName()) {
+                    return false;
+                }
+                if (states[index].getHighlighted() !== otherStates[index].getHighlighted()) {
+                    return false;
+                }
+                if (states[index].getSystemAction() !== otherStates[index].getSystemAction()) {
+                    return false;
+                }
+                if (states[index].getSoftButton() === null && otherStates[index].getSoftButton() === null) {
+                    continue;
+                }
+                if (states[index].getSoftButton().getParameters() !== otherStates[index].getSoftButton().getParameters()) {
```

Comment:
This is a reference equality check and not a deep equality check, so comparing objects like this will fail:

```
        const m1 = new SDL.manager.screen.utils.SoftButtonObject('hello', new SDL.manager.screen.utils.SoftButtonState('hi', 'ok'), 'hi')

        const m2 = new SDL.manager.screen.utils.SoftButtonObject('hello', new SDL.manager.screen.utils.SoftButtonState('hi', 'ok'), 'hi')
        console.log(m1.equals(m2));
```

You _may_ be able to get away with doing a stringified JSON equality check since the soft button's properties will guarantee to be set in a specific order (via the SoftButtonState's constructor method) so that the resulting object is the same. Testing would be needed:
`JSON.stringify(states[index].getSoftButton().getParameters()) !== JSON.stringify(otherStates[index].getSoftButton().getParameters())`

Reply:
Stringifying the parameters wouldn't work here because the Image param is a Struct which would need to be cloned itself making the logic a little messier. I reverted this back to the way it was before.